File: datastructures/src/main/java/com/javaground/datastructures/trees/trie/TrieNode.java
package com.javaground.datastructures.trees.trie;

import java.util.HashMap;

/**
 * Represents a single node in a Trie data structure.
 * A TrieNode stores a flag to denote the end of a word
 * and keeps track of its children nodes using a map.
 */
public class TrieNode {
    public boolean isEnd;
    public HashMap<Character, TrieNode> children;

    public TrieNode() {
        this.isEnd = false;
        this.children  = new HashMap<>();
    }
}

File: datastructures/src/main/kotlin/com/kotlinground/datastructures/trees/trie/TrieNode.kt
package com.kotlinground.datastructures.trees.trie

/**
 * TrieNode represents a node in a Trie
 * @param children [HashMap] is a hashmap of character to node
 * @param isEnd [Boolean] whether this the end of a word
 */
data class TrieNode(
    val children: HashMap<Char, TrieNode> = hashMapOf(),
    var isEnd: Boolean = false
)
